client/index.ts

```diff
@@ -78,25 +78,77 @@ class OAuth2Client {
   }
 
   private async getJWKS(): Promise<JWKSDT> {
+    if (this.jwks) {
+      return Promise.resolve(this.jwks);
+    } else {
+      return await this.fetchJWKS();
+    }
+  }
+
+  private async fetchJWKS(): Promise<JWKSDT> {
     const openIDConfiguration = await this.getOpenIDConfiguration();
 
-    if (!this.openIDConfiguration.jwks_uri) {
+    if (!openIDConfiguration.jwks_uri) {
       throw new MissingJWKSURI("Missing JWKS URI for RS256 encoded JWT");
     }
 
-    if (this.jwks) {
-      return Promise.resolve(this.jwks);
-    } else {
-      return await this.fetch(openIDConfiguration.jwks_uri)
-        .then((response) => response.json())
-        .then((jwks) => {
-          this.jwks = jwks;
-          return jwks;
-        })
-        .catch((error) => {
-          throw error;
-        });
+    return await this.fetch(openIDConfiguration.jwks_uri)
+      .then((response) => response.json())
+      .then((jwks) => {
+        this.jwks = jwks;
+        return jwks;
+      })
+      .catch((error) => {
+        throw error;
+      });
+  }
+
+  private async getPublicKey(jwks: JWKSDT, kid: string): Promise<string> {
+    const validKey = jwks.keys.find((keyObject) => keyObject.kid === kid);
+
+    if (validKey) {
+      const { e, n } = validKey;
+      const publicKey = rsaPublicKeyToPEM(n, e);
+      return publicKey;
     }
```

Comment:
Why not put the retry in there?

Reply:
To keep things separated I think, since the checking of kid in JWKS and key trasnformation in PEM format is a common thing between 1st try and other ones, I thought it would be better. 

```typescript
  private async getPublicKey(
    jwks: JWKSDT,
    kid: string,
    accessToken: string,
  ): Promise<string> {
    const validKey = jwks.keys.find((keyObject) => keyObject.kid === kid);

    if (validKey) {
      const { e, n } = validKey;
      const publicKey = rsaPublicKeyToPEM(n, e);
      return await this.verifyRS256(accessToken, publicKey);
    } else {
      const updatedJwks = await this.fetchJWKS();
      const validKey = updatedJwks.keys.find(
        (keyObject) => keyObject.kid === kid,
      );

      if (validKey) {
        const { e, n } = validKey;
        const publicKey = rsaPublicKeyToPEM(n, e);
        return await this.verifyRS256(accessToken, publicKey);
      } else {
        throw new InvalidKeyIDRS256(
          "Invalid key ID (kid) for RS256 encoded JWT",
        );
      }
    }
  }
```

Would you prefer this function to do all the job like this ? (I can't find an approriate new naming for its task 😆 )

Reply:
I didn't mean to do the `verify` in there but just add the refetch in the `getPublicKey` function.

Here's a counter proposal to avoid a bit of duplication:

```ts
private async getPublicKey(
    jwks: JWKSDT,
    kid: string,
  ): Promise<string> {
    let validKey = jwks.keys.find((keyObject) => keyObject.kid === kid);
    If (!validKey) {
      const updatedJwks = await this.fetchJWKS();
      validKey = updatedJwks.keys.find(
        (keyObject) => keyObject.kid === kid,
      );
    }

    if (validKey) {
      const { e, n } = validKey;
      const publicKey = rsaPublicKeyToPEM(n, e);
      return publicKey;
    } else {
        throw new InvalidKeyIDRS256(
          "Invalid key ID (kid) for RS256 encoded JWT",
        );
    }
  }
  ```

Reply:
Ah alright, I get it now. And I like it, let's do that. 👍